### ISSUE TITLE
`@angular/common/http` request throws an error that cannot be unit tested properly

### ISSUE THREAD

**Issue by peterbakonyi05** [angular/angular#19289]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x ] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
When `responseType` is `json` and the body of the response is a string (e.g.: url `http://domain.com/url-of-the-created-resource`), there is an error because [JSON.parse](https://github.com/angular/angular/blob/4.4.3/packages/common/http/src/xhr.ts#L187) fails. Previously using `res.json()` with `@angular/http` this did not throw an error but returned the `string` as is. We had unit tests for this but `TestRequest` behaves differently and it does not throw an error so it is impossible to test this behavior using `@angular/common/http/testing`.

## Expected behavior
A: `TestRequest` should behave the same way and throw an error in this case
B: Do not throw an error and have the same behavior as `@angular/http` had

## Minimal reproduction of the problem with instructions
```ts
this.http.request(method, url, {
    observe: "response",
    responseType: "json"
});
```

If body of the response is a string like "https://github.com" it throws an error with real XHR. However it does not throw using `TestRequest.flush`

## What is the motivation / use case for changing the behavior?
To improve reliability of unit tests or fix a potential issue.

## Environment
Angular version: 4.X.X

Browser:
<pre><code>
- [X ] Chrome (desktop) version Latest
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
Others:

</code></pre>

**Comment by alxhub** [comment]
This is true, and also by design. `HttpClientTestingModule` actually works at a level of abstraction above the XHR layer - that is, above serialized requests. That is, the argument you pass to `flush` is the _deserialized_ body already, which as you mention assumes successful deserialization.

You're absolutely right that this is a missing piece of the testing API, and it's something we should provide.

**Comment by PapaNappa** [comment]
In my use-case, the component is loading binary file contents using `httpClient.get(url, { responseType: 'arraybuffer' })`.
In case of an error (e.g. 404), a ProblemDetails JSON is returned.

The workaround for me is to use the clunky
```ts
const buffer = new TextEncoder().encode('{"Detail": "not found"}').buffer;
request.flush(buffer, {status: 404, statusText: 'Not Found'})
```
